### PR TITLE
Option to space out calculations

### PR DESCRIPTION
**Pull request by Madrawn**
I noticed on a new map with only one pawn that enabling or disabling free will on the pawn was the difference between 300 ticks per second and 600 ticks per second while the pawn slept.
Analyzing with Dubs Performance Analyzer further I saw that the FreeWill_MapComponent took 90% of time spent in "Map Component Tick" and 99% of that was "System:Action.Invoke". So I looked into your code and decided that it won't need to spend time to calculate and set priorities every tick and made it so it only execute the tick action every X ticks.

This simple "fix" works fine for me as I'm playing it right now.

Basically makes it so it doesn't calculate every tick but skips some between calculations Integrated into the options menu
defaults to 20 ticks pause

### CODE REVIEW

**Comment by paul-freeman** [comment]
My understanding was that ticks were fixed at [60 per second](https://rimworldwiki.com/wiki/Time#:~:text=Time%20in%20RimWorld%20passes%20as,to%20both%20of%20these%20numbers.). So, I think doing this will just space out the calculations. This could be fine, but I'll need to test it. Unless maybe you can explain how you were gettng 300 - 600 ticks per second. Is that maybe a mod?

In general, I like the idea, so I'll look to do something with this PR.

**Comment by Madrawn** [comment]
_Time in RimWorld passes as a series of ticks. On normal game speed, there are 60 such ticks to a real second; there are always 2,500 ticks to one in-game hour. The play speed controls add multipliers to both of these numbers._

60 ticks per second only applies to 1x speed.
So when running at max speed the game tries to reach 360 ticks per second and when everyone is asleep it even goes as high as 720 TPS

I'd recommend you look into [DubsPerformanceAnalyzer](https://steamcommunity.com/sharedfiles/filedetails/?id=2038874626) it gets you an ingame performance analyzer and shows the TPS on screen. It's really mighty and lets you check out the call performance down to the system methods. It's how I found the spot inside your code to space out. (rightclick a call to step into a call to see the calls inside; took me way to long to find that feature)

**Review by paul-freeman — CHANGES_REQUESTED** [review]
If you knock the default ticks down, I'll merge this in and put a build up on the unstable version of the mod. Thanks for the contribution!

**Comment by paul-freeman** [comment]
Thanks. I didn't know it went up that high. I have used the analyzer before when I needed to speed up the mod.

To give you some more background, it doesn't calculate the priorities every tick. It does one worktype for one pawn each tick. This prevents it from getting slower as you get more pawns (or if you add mods with more worktypes) and also greatly reduces the amount of calculations per tick.

So, if you space everything out by 20 ticks, then if you have a large base with say 20 pawns and 15 worktypes, you would only be updating the colony priorities every 6000 ticks, which is a couple in-game hours between updates. We could probably set the default spacing to 3 or 4 ticks, but 20 is probably going to impact some players.

One thing that seems obvious would be to just skip the calculations while they are sleeping (or downed or whatever). This would essentially disable the mod at night at probably give the same results. That combined with some fine tuning of the worst parts of the AI would probably make a big difference.

File: FreeWill_ModSettings.cs [hunk 1]
```diff
@@ -8,6 +8,7 @@ namespace FreeWill
     public class FreeWill_ModSettings : ModSettings
     {
         // mod default settings
+        const int SkipCalcTicksDefault = 20;
```
[the remark this answers is not in the document]
Comment (Madrawn) [reply FreeWill_ModSettings.cs:11]: How frequent do you recon that the work priorities should be recalculated? 4 seems low as it still takes up around 40% of all time spent executing map components.
image 

I'd go at least for 10 as a default. Testing that locally it leaves the mod 3% time taken.

----------
_Some math for me to think:_

_The first 8 actions seem to be the most cpu intensive.  
Without pause ticks the "getMapComponentTickAction" repeats every `pawn * 20 + 8` ticks.  
The default AI aims for 12-15 pawns. Let's say 10.   
So repeat happens every 20 - 200 ticks (1 - 10 pawns), which is 3 times per second or every 3 seconds.  
Divide that by 4 to consider 4x speed so = 12 times per second or 0.75 per second._

-------------
I would argue that one would need to update the priorities every 5 seconds at most, 4 times speed, as that is still longer than most active tasks take and the new priorities will only matter after a pawn finishes his current task. That would mean a default between 7 (10 pawns) or 62 (1 pawn), depending on number of pawns.

Honestly, the problem I see is that the performance get's better the more pawns you have, as the expensive calculations are farther apart, which makes it difficult to decide on a number.

**Review by Madrawn — COMMENTED** [review]
See my comment about the default-value

**Comment by paul-freeman** [comment]
Closing due to new 1.6 changes in RimWorld
